src/server/game/BlackMarket/BlackMarketMgr.cpp

```diff
@@ -0,0 +1,488 @@
+
+    if (!item)
+        return;
+
+    if (templ->Item.ItemBonus)
+        for (int32 bonusList : templ->Item.ItemBonus->BonusListIDs)
+            item->AddBonuses(bonusList);
+
+    item->SetOwnerGUID(bidderGuid);
+
+    item->SaveToDB(trans);
+
+    // Log trade
+    if (logGmTrade)
+    {
+        sLog->outCommand(bidderAccId, "GM %s (Account: %u) won item in blackmarket auction: %s (Entry: %u Count: %u) and payed gold : %u.",
+            bidderName.c_str(), bidderAccId, item->GetTemplate()->GetDefaultLocaleName(), item->GetEntry(), item->GetCount(), entry->CurrentBid / GOLD);
+    }
+
+    if (bidder)
+    {
+        bidder->GetSession()->SendBlackMarketWonNotification(entry, item);
+    }
+
+    MailDraft(entry->BuildAuctionMailSubject(BMAH_AUCTION_WON), entry->BuildAuctionMailBody())
+        .AddItem(item)
+        .SendMailTo(trans, MailReceiver(bidder, entry->Bidder), entry, MAIL_CHECK_MASK_COPIED);
+
+    entry->MailSent();
+}
+
+void BlackMarketMgr::SendAuctionOutbidMail(BlackMarketEntry* entry, SQLTransaction& trans)
+{
+    ObjectGuid oldBidder_guid = ObjectGuid::Create<HighGuid::Player>(entry->Bidder);
+    Player* oldBidder = ObjectAccessor::FindConnectedPlayer(oldBidder_guid);
+
+    uint32 oldBidder_accId = 0;
+    if (!oldBidder)
+        oldBidder_accId = ObjectMgr::GetPlayerAccountIdByGUID(oldBidder_guid);
+
+    // old bidder exist
+    if (!oldBidder && !oldBidder_accId)
+        return;
+
+    if (oldBidder)
+        oldBidder->GetSession()->SendBlackMarketOutbidNotification(entry->GetTemplate());
+
+    MailDraft(entry->BuildAuctionMailSubject(BMAH_AUCTION_OUTBID), entry->BuildAuctionMailBody())
+        .AddMoney(entry->CurrentBid)
+        .SendMailTo(trans, MailReceiver(oldBidder, entry->Bidder), entry, MAIL_CHECK_MASK_COPIED);
+}
+
+BlackMarketEntry* BlackMarketMgr::GetAuctionByID(int32 marketId) const
+{
+    BlackMarketEntryMap::const_iterator itr = mAuctions.find(marketId);
+    if (itr != mAuctions.end())
+        return itr->second;
+
+    return nullptr;
+}
+
+BlackMarketTemplate* BlackMarketMgr::GetTemplateByID(int32 marketId) const
+{
+    BlackMarketTemplateMap::const_iterator itr = mTemplates.find(marketId);
+    if (itr != mTemplates.end())
+        return itr->second;
+
+    return nullptr;
+}
+
+bool BlackMarketTemplate::LoadFromDB(Field* fields)
+{
+    MarketID = fields[0].GetInt32();
+    SellerNPC = fields[1].GetInt32();
+    Item.ItemID = fields[2].GetUInt32();
+    Quantity = fields[3].GetInt32();
+    MinBid = fields[4].GetUInt64();
+    Duration = static_cast<time_t>(fields[5].GetUInt32());
+    Chance = fields[6].GetFloat();
+
+    Tokenizer bonusListIDsTok(fields[7].GetString(), ' ');
+    std::vector<int32> bonusListIDs;
+    for (char const* token : bonusListIDsTok)
+    {
+        int32 bonusListID = atol(token);
+        bonusListIDs.push_back(bonusListID);
+    }
+
+    if (!bonusListIDs.empty())
+    {
+        Item.ItemBonus.emplace();
```

Comment:
no member named 'emplace' in
      'boost::optional<WorldPackets::Item::ItemBonusInstanceData>'
        Item.ItemBonus.emplace();